ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs

```diff
@@ -27,26 +28,29 @@ type internal TypeProvidersCache() =
                 oldTp.Dispose()
                 addTypeProvider envKey tp
             | false, _ ->
-                assemblyCache.Add(fullName, tp)
-                proxyTypeProvidersPerId.Add(tp.EntityId, tp)
+                assemblyCache.TryAdd(fullName, tp) |> ignore
+                proxyTypeProvidersPerId.TryAdd(tp.EntityId, tp) |> ignore
                 tp.Disposed.Add(fun _ -> removeTypeProvider envKey tp)
         | false, _ ->
-            typeProvidersPerAssembly.Add(envKey, Dictionary())
+            typeProvidersPerAssembly.TryAdd(envKey, ConcurrentDictionary()) |> ignore
             addTypeProvider envKey tp
 
     and removeTypeProvider envKey (tp: IProxyTypeProvider) =
         // Removes types in a unified manner, may also be disposed by FCS.
-        typeProvidersPerAssembly.[envKey].Remove(tp.GetDisplayName(true)) |> ignore
-        proxyTypeProvidersPerId.Remove(tp.EntityId) |> ignore
+        typeProvidersPerAssembly.[envKey].TryRemove(tp.GetDisplayName(true)) |> ignore
+        proxyTypeProvidersPerId.TryRemove(tp.EntityId) |> ignore
 
         if typeProvidersPerAssembly.[envKey].Count = 0 then
-            typeProvidersPerAssembly.Remove(envKey) |> ignore
+            typeProvidersPerAssembly.TryRemove(envKey) |> ignore
 
     member x.Add(envKey, tp) =
         addTypeProvider envKey tp
 
     member x.Get(id) =
-        proxyTypeProvidersPerId.[id]
+        let hasValue = SpinWait.SpinUntil((fun () -> proxyTypeProvidersPerId.ContainsKey id), 15_000)
```

Comment:
Where does `15_000` literal comes from?

Reply:
this is a fairly random number (15 seconds). It is assumed that the data will appear in the cache quickly enough, but for an unforeseen situation, we can wait a little longer.

Reply:
do you think 15 seconds is a lot?

Reply:
No, it's choosing the time arbitrary seems very suspicious to me.